standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -5104,14 +5103,34 @@ public boolean drop_partition(final String db_name, final String tbl_name,
         null);
   }
 
-  private static class PathAndPartValSize {
-    PathAndPartValSize(Path path, int partValSize) {
-      this.path = path;
-      this.partValSize = partValSize;
+    /** Stores a path and its size. */
+    private static class PathAndPartValSize implements Comparable<PathAndPartValSize> {
+
+      public Path path;
+      int partValSize;
+
+      public PathAndPartValSize(Path path, int partValSize) {
+        this.path = path;
+        this.partValSize = partValSize;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        if (o == this) {
+          return true;
+        }
+        if (!(o instanceof PathAndPartValSize)) {
+          return false;
+        }
+        return path.equals(((PathAndPartValSize) o).path);
```

Comment:
why we only compare `path` but not `partValSize`?

Reply:
Nice catch. It is a bug. The current code actually didn't correctly implement the HashSet, it just used the hashcode of the object but not (path, depth) pair.